### PR TITLE
feat(cone/caratheodory): add caratheodory theorem for convex cones

### PR DESCRIPTION
**Pull request by apurvanakade**
Carathéodory's theorem states that if `x` is in the convex cone generated by a set `s`, then `x` is a convex combination of linearly independent elements of `s`. This is the replacement of the usual statement of vector spaces having bases for the setting of convex cones.

This is needed for proving duality theorems for linear programs.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

- [x] depents on: #8878

### CODE REVIEW

File: Mathlib/Analysis/Convex/Cone/Caratheodory.lean [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+end Caratheodory
+
+variable {s : Set E}
+
+/-- **Carathéodory's convexity theorem for convex cones** -/
+
+theorem toPointedCone_eq_union : (s.toPointedCone 𝕜 : Set E) =
+    ⋃ (t : Finset E) (_ : ↑t ⊆ s) (_ : LinearIndependent 𝕜 ((↑) : t → E)),
+      (SetLike.coe $ toPointedCone 𝕜 t) := by
```
Comment (AntoineChambert-Loir) [inline Mathlib/Analysis/Convex/Cone/Caratheodory.lean:214]: I have the impression that in lean4, the notation $ has been replaced by <| …

File: Mathlib/Analysis/Convex/Cone/Caratheodory.lean [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    simp only [exists_prop, Set.mem_iUnion]
+    exact ⟨Caratheodory.minCardFinsetOfMemtoPointedCone hx,
+      Caratheodory.minCardFinsetOftoPointedCone_subseteq hx,
+      Caratheodory.linearIndependent_minCardFinsetOfMemtoPointedCone hx,
+      Caratheodory.mem_minCardFinsetOfMemtoPointedCone hx⟩
+  · iterate 3 convert Set.iUnion_subset _; intro
+    exact Submodule.span_mono ‹_›
+
+/-- A more explicit version of `toPointedCone_eq_union`. -/
+theorem eq_pos_convex_span_of_mem_toPointedCone {x : E} (hx : x ∈ s.toPointedCone 𝕜) :
```
Comment (AntoineChambert-Loir) [inline Mathlib/Analysis/Convex/Cone/Caratheodory.lean:226]: can you recast this as an iff theorem?

File: Mathlib/Analysis/Convex/Cone/Caratheodory.lean (hunk 1 above)
Comment (apurvanakade) [reply Mathlib/Analysis/Convex/Cone/Caratheodory.lean]: Oops. Fixed, thanks!

**Comment by ocfnash** [comment]
It looks like the repeats all the work used to establish [convexHull_eq_union](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Convex/Caratheodory.html#convexHull_eq_union) except now in this more restricted setting. Wouldn't it be possible to use `convexHull_eq_union` to obtain this result. E.g., is this true?
```lean
lemma toPointedCone_eq_smul_convexHull :
    s.toPointedCone 𝕜 = {y | ∃ x ∈ convexHull 𝕜 (s ∪ {0}), ∃ t : 𝕜, 0 ≤ t ∧ t • x = y} :=
  sorry
```

File: Mathlib/Analysis/Convex/Cone/Caratheodory.lean [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+- `toPointedCone_eq_union`: Carathéodory's convexity theorem for convex cones.
+- `mem_toPointedCone_iff_eq_pos_convex_span`: A more explicit version of `toPointedCone_eq_union`.
+
+-/
+
+
+universe u
+
+open Set Finset
```
Comment (madvorak) [inline Mathlib/Analysis/Convex/Cone/Caratheodory.lean:30]: Why are these two on the same line but `BigOperators` on a separate line?

Comment (apurvanakade) [reply Mathlib/Analysis/Convex/Cone/Caratheodory.lean:30]: Oops, I'll fix this. Thanks!

**Comment by ocfnash** [comment]
> > It looks like this repeats all the work used to establish [convexHull_eq_union](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Convex/Caratheodory.html#convexHull_eq_union) except now in this more restricted setting. Wouldn't it be possible to use `convexHull_eq_union` to obtain this result. E.g., is this true?
> > ```lean
> > lemma toPointedCone_eq_smul_convexHull :
> >     s.toPointedCone 𝕜 = {y | ∃ x ∈ convexHull 𝕜 (s ∪ {0}), ∃ t : 𝕜, 0 ≤ t ∧ t • x = y} :=
> >   sorry
> > ```
> 
> This statement is true but you still need a way to go from `AffineIndependent` to `LinearlyIndependent`. The fact that the difference vectors from a single point in an `AffineIndependent` set are `LinearlyIndependent` is not sufficient here. I could not figure out to cross this bridge.

Are you saying that [https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/AffineSpace/Independent.html#affineIndependent_iff_linearIndependent_vsub](affineIndependent_iff_linearIndependent_vsub) is not sufficient?

If so, can you provide a formal statement of what you need?

**Comment by apurvanakade** [comment]
Ok, I found a reference online that uses one to prove the other - although it is in the other direction. 
I'll try to see if I can flip the proof: https://www.math.cuhk.edu.hk/course_builder/1819/math4230/caratheodory.pdf

**Comment by AntoineChambert-Loir** [comment]
Shameless self-advertising: Corollary 1.2.2 in
https://webusers.imj-prg.fr/~antoine.chambert-loir/enseignement/2020-21/gt/toptrop.pdf

**Comment by apurvanakade** [comment]
> Shameless self-advertising: Corollary 1.2.2 in https://webusers.imj-prg.fr/~antoine.chambert-loir/enseignement/2020-21/gt/toptrop.pdf

Thank you! I'll open up a thread on Zulip to get help with formalizing this.

**Comment by apurvanakade** [comment]
@AntoineChambert-Loir 

I've looked at the proof in your notes and while it does connect the two Caratheodory theorems it is much much longer. The easier direction in your proof is to prove the Cone version version and then use it to prove the Conic Hull version. Here we are trying to do the other direction which goes through several intermediate steps, each of which will be a big Mathlib PR. I won't have time to do this right now.

@ocfnash Let me know what you suggest. There is no shortcut to go from Convex Hull to Cones (there is one to go the other way around).

**Comment by ocfnash** [comment]
> AntoineChambert-Loir
> 
> I've looked at the proof in your notes and while it does connect the two Caratheodory theorems it is much much longer. The easier direction in your proof is to prove the Cone version first and then use it to prove the Conic Hull version. Here we are trying to do the other direction which goes through several intermediate steps, each of which will be a big Mathlib PR. I won't have time to do this right now.
> 
> ocfnash Let me know what you suggest. There is no shortcut to go from Convex Hull to Cones (there is one to go the other way around).

I suggest you start by proving the lemma whose statement I provided [here](https://github.com/leanprover-community/mathlib4/pull/9251#issuecomment-1891796850) --- it would be a useful contribution in its own right and it looks to me like it should be easy to obtain `toPointedCone_eq_union` from `convexHull_eq_union` using it.

I find it very hard to believe there is no shortcut between these two results, but even if there is something I have overlooked we prefer to put in the extra work and avoid the sort of code duplication currently proposed. Mathlib works because we all put in the extra effort.

**Comment by apurvanakade** [comment]
Sounds good. I'll work on these smaller lemmas first.

**Comment by YaelDillies** [comment]
(since Apurva has contributed extensively to mathlib3)

**Comment by jcommelin** [comment]
@apurvanakade Are you planning to continue working on this PR? If not, please feel free to close it.

**Comment by apurvanakade** [comment]
I haven't had time to work on Mathlib of late, sorry. I'll close the PR.

**Comment by YaelDillies** [comment]
Feel free to reopen it when you have time again!

**Pull request closed without merging**